torchx/schedulers/lsf_scheduler.py

```diff
@@ -0,0 +1,513 @@
+)
+from torchx.schedulers.ids import make_unique
+from torchx.schedulers.local_scheduler import LogIterator
+from torchx.specs import (
+    AppDef,
+    AppState,
+    BindMount,
+    DeviceMount,
+    VolumeMount,
+    macros,
+    NONE,
+    ReplicaStatus,
+    Role,
+    RoleStatus,
+    runopts,
+)
+from torchx.workspace.dir_workspace import DirWorkspace
+from typing_extensions import TypedDict
+
+JOB_STATE: Dict[str, AppState] = {
+    "DONE": AppState.SUCCEEDED,
+    "EXIT": AppState.FAILED,
+    "PEND": AppState.PENDING,
+    "RUN": AppState.RUNNING,
+    "PSUSP": AppState.PENDING,
+    "USUSP": AppState.PENDING,
+    "SSUSP": AppState.PENDING,
+}
+
+def get_job_state(state_str: str, exit_code: str) -> AppState:
+    state = AppState.UNKNOWN
+    if state_str in JOB_STATE.keys():
+        state = JOB_STATE[state_str]
+    if state == AppState.FAILED and exit_code == "130": # likely SIGINT
+        state = AppState.CANCELLED
+    return state
+
+class LsfOpts(TypedDict, total=False):
+    lsf_queue: Optional[str]
+    jobdir: Optional[str]   # NOTE: *job_dir* cannot be used. somehow it overwrites --image flag (bug?). so use *jobdir* (without underscore, _)
+    runtime: str
+    container_workdir: Optional[str]
+    host_network: Optional[bool]
+    shm_size: Optional[str]
+
+def get_docker_command(job_name: str, role: Role, cfg: LsfOpts) -> str:
+    cmds = ["docker", "run", f"--name={job_name}"]
+    for mount in role.mounts:
+        rw = "rw"
+        if mount.read_only:
+            rw = "ro"
+        if isinstance(mount, BindMount):
+            cmds += ["-v", f"{mount.src_path}:{mount.dst_path}:{rw}"]
+        elif isinstance(mount, VolumeMount):
+            cmds += ["--mount", f"\"type=volume,src={mount.src_path}:dst={mount.dst_path},{rw}\""]
+        elif isinstance(mount, DeviceMount):
+            cmds += [f"--device={mount.src_path}:{mount.dst_path}:{mount.permissions}"]
+        else:
+            raise Exception(f"Unknown mount type: {mount}")
+    container_workdir = cfg.get("container_workdir")
+    if container_workdir:
+        cmds += ["-w", container_workdir]
+    host_network = cfg.get("host_network")
+    if host_network:
+        cmds += ["--net=host", "--ipc=host"]
+    else:
+        for name, port in role.port_map.items():
+            cmds += ["-p", str(port)]
+    shm_size = cfg.get("shm_size")
+    if shm_size:
+        cmds += [f"--shm-size={shm_size}"]
+    for key, value in dict(role.env).items():
+        cmds += ["-e", f"{key}=\"{value}\""]
+
+    resource = role.resource
+    if resource != NONE:
+        if resource.cpu > 0:
+            cmds += [f"--cpus={str(resource.cpu)}"]
+        if resource.memMB > 0:
+            cmds += [f"--memory={str(resource.memMB * 1024 * 1024)}"]
+        if resource.gpu > 0:
+            cmds += [f"--gpus", "all"]
+    cmds += ["--entrypoint", role.entrypoint, "--rm", role.image] + [shlex.quote(arg) for arg in role.args]
+    return " ".join(cmds).replace("$", "\\$")
```

Comment:
need to use something like shlex.join -- just using " ".join isn't safe from both a correctness and security standpoint

applies below as well